src/systems/renderer.js

```diff
@@ -83,3 +96,59 @@ module.exports.System = registerSystem('renderer', {
     }
   }
 });
+
+// Custom A-Frame sort functions.
+// Variations of Three.js default sort orders here:
+// https://github.com/mrdoob/three.js/blob/ebbaecf9acacf259ea9abdcba7b6fb25cfcea2ab/src/renderers/webgl/WebGLRenderLists.js#L1
+// See: https://github.com/aframevr/aframe/issues/5332
+
+// Default sort for opaque objects:
+// - respect groupOrder & renderOrder settings
+// - sort front-to-back by z-depth from camera (this should minimize overdraw)
+// - otherwise leave objects in default order (object tree order)
+
+function sortSpatialFrontToBack (a, b) {
+  if (a.groupOrder !== b.groupOrder) {
+    return a.groupOrder - b.groupOrder;
+  } else if (a.renderOrder !== b.renderOrder) {
+    return a.renderOrder - b.renderOrder;
+  } else if (a.z !== b.z) {
+    return a.z - b.z;
+  } else {
+    return 0;
```

Comment:
I think this statement might be redundant. If one does `a.z - b.z` if they are equal they will return 0 as well.

Reply:
similarly:

```
return (a.groupOrder - b.groupOrder) || (a.renderOrder - b.renderOrder) || (a.z - b.z);

```

Reply:
Wouldn't it make sense to retain the ID tie-breaker (`return a.id - b.id`) used by Three.js as well? That should ensure that, everything else being equal, the order is consistent.

As for compacting it to a single expression, I personally prefer the current more verbose version. It's easier to see what's going on and it resembles the referenced original Three.js code, making it easy to compare. I did experiment with some very crude benchmarking and while both perform about the same, a significant difference forms once some of the properties are (mostly) consistent. It seems the if-else chain allows for better branch-prediction:

- [0 groupOrder, 1% renderOrder, random Z](https://perf.link/#eyJpZCI6Im04NHh5amJxOXE2IiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6ImNvbnN0IGFycmF5ID0gbmV3IEFycmF5KDEwMDAwKTtcbmZvcihsZXQgaSA9IDA7IGkgPCBhcnJheS5sZW5ndGg7IGkrKykge1xuXHRhcnJheVtpXSA9IHtcblx0XHRncm91cE9yZGVyOiAwLFxuXHRcdHJlbmRlck9yZGVyOiBNYXRoLnJhbmRvbSgpIDwgMC4wMSA%2FIE1hdGguZmxvb3IoTWF0aC5yYW5kb20oKSoxMDApIDogMCxcblx0XHR6OiBNYXRoLnJhbmRvbSgpICogMTAwMDAwIC0gNTAwMDAsXG5cdH07XG59IiwidGVzdHMiOlt7Im5hbWUiOiJGaW5kIGl0ZW0gMTAwIiwiY29kZSI6ImFycmF5LnNvcnQoKGEsYikgPT4ge1xuXHRpZihhLmdyb3VwT3JkZXIgIT09IGIuZ3JvdXBPcmRlcikge1xuXHRcdHJldHVybiBhLmdyb3VwT3JkZXIgLSBiLmdyb3VwT3JkZXI7XG5cdH0gZWxzZSBpZihhLnJlbmRlck9yZGVyICE9PSBiLmdyb3VwT3JkZXIpIHtcblx0XHRyZXR1cm4gYS5yZW5kZXJPcmRlciAtIGIucmVuZGVyT3JkZXI7XG5cdH0gZWxzZSB7XG5cdFx0cmV0dXJuIGEueiAtIGIuejtcblx0fVxufSkiLCJydW5zIjpbMjI1MCwyNjI1LDIwMDAsMTYyNSwyNzUwLDIyNTAsMjI1MCwyMTI1LDIxMjUsMjAwMCwyMzc1LDIwMDAsMjYyNSwyMjUwLDE1MDAsMjUwMCwyMjUwLDE4NzUsMjEyNSwxODc1LDIxMjUsMjI1MCwxODc1LDE3NTAsMjAwMCwyMTI1LDIzNzUsMjc1MCwxODc1LDE4NzUsMTUwMCwxODc1LDE3NTAsMTc1MCwyNzUwLDIyNTAsMTc1MCwyMTI1LDE4NzUsMTg3NSwxODc1LDIwMDAsMTg3NSwyMTI1LDIyNTAsMjAwMCwyNTAwLDIyNTAsMjM3NSwyMDAwLDI1MDAsMTg3NSwyMDAwLDE3NTAsMjI1MCwyMTI1LDIyNTAsMjEyNSwyMDAwLDIxMjUsMjAwMCwyMjUwLDIyNTAsMjEyNSw1MDAsMTc1MCwxNjI1LDI1MDAsMjI1MCwyMjUwLDIwMDAsMjM3NSwyMzc1LDIyNTAsMjUwMCwyMTI1LDE3NTAsMjYyNSwyMzc1LDE4NzUsMTYyNSwyMDAwLDE4NzUsMjUwMCwxMzc1LDIwMDAsMTg3NSwyMzc1LDIwMDAsMjM3NSwxODc1LDE4NzUsMTYyNSwxNzUwLDEzNzUsMTg3NSwyMDAwLDE2MjUsMTYyNSwyMDAwXSwib3BzIjoyMDU1fSx7Im5hbWUiOiJGaW5kIGl0ZW0gMjAwIiwiY29kZSI6ImFycmF5LnNvcnQoKGEsYikgPT4gKGEuZ3JvdXBPcmRlciAtIGIuZ3JvdXBPcmRlcikgfHwgKGEucmVuZGVyT3JkZXIgLSBiLnJlbmRlck9yZGVyKSB8fCAoYS56IC0gYi56KSkiLCJydW5zIjpbMTEyNSwxMTI1LDEzNzUsMTc1MCwxNzUwLDE2MjUsMjAwMCwxNTAwLDE2MjUsMTEyNSwzNzUsMTUwMCwxNTAwLDEyNTAsMTAwMCwxMjUwLDg3NSwyMDAwLDEwMDAsMTM3NSwxMjUwLDExMjUsMTUwMCwxMzc1LDE1MDAsMTM3NSwyMTI1LDE4NzUsMTc1MCwxNzUwLDE2MjUsMTM3NSwyMDAwLDE3NTAsMTg3NSwxNTAwLDE1MDAsMTM3NSwxNjI1LDEyNTAsMTUwMCwxNjI1LDEwMDAsMTUwMCwyMDAwLDE4NzUsMTc1MCwxNTAwLDE3NTAsMTM3NSwxNTAwLDEwMDAsMTYyNSwxNzUwLDEzNzUsMTYyNSwxMzc1LDExMjUsMTI1MCwxNjI1LDEzNzUsMTUwMCwxNjI1LDIwMDAsMTc1MCwxNzUwLDEzNzUsMTc1MCwxODc1LDE1MDAsMTYyNSwxNjI1LDEyNTAsMTUwMCwxNjI1LDE2MjUsMTUwMCwxNjI1LDExMjUsMTEyNSwxNjI1LDE1MDAsMTI1MCwxMzc1LDE3NTAsMTc1MCwxNTAwLDE1MDAsMTI1MCwxMDAwLDE1MDAsMTYyNSwxMDAwLDE1MDAsMTEyNSwxMTI1LDEzNzUsMjAwMCwxMzc1LDEyNTBdLCJvcHMiOjE0ODJ9XSwidXBkYXRlZCI6IjIwMjMtMDgtMDhUMDc6NDE6NTMuNTI4WiJ9): If-else chain faster on Chrome/Quest 2, slower on Firefox
- [random groupOrder, random renderOrder, random z](https://perf.link/#eyJpZCI6IjcyMm44ZWo1YjQzIiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6ImNvbnN0IGFycmF5ID0gbmV3IEFycmF5KDEwMDAwKTtcbmZvcihsZXQgaSA9IDA7IGkgPCBhcnJheS5sZW5ndGg7IGkrKykge1xuXHRhcnJheVtpXSA9IHtcblx0XHRncm91cE9yZGVyOiBNYXRoLmZsb29yKE1hdGgucmFuZG9tKCkgKiAxMDAwMDApLFxuXHRcdHJlbmRlck9yZGVyOiBNYXRoLmZsb29yKE1hdGgucmFuZG9tKCkgKiAxMDAwMDApLFxuXHRcdHo6IE1hdGgucmFuZG9tKCkgKiAxMDAwMDAgLSA1MDAwMCxcblx0fTtcbn0iLCJ0ZXN0cyI6W3sibmFtZSI6IkZpbmQgaXRlbSAxMDAiLCJjb2RlIjoiYXJyYXkuc29ydCgoYSxiKSA9PiB7XG5cdGlmKGEuZ3JvdXBPcmRlciAhPT0gYi5ncm91cE9yZGVyKSB7XG5cdFx0cmV0dXJuIGEuZ3JvdXBPcmRlciAtIGIuZ3JvdXBPcmRlcjtcblx0fSBlbHNlIGlmKGEucmVuZGVyT3JkZXIgIT09IGIuZ3JvdXBPcmRlcikge1xuXHRcdHJldHVybiBhLnJlbmRlck9yZGVyIC0gYi5yZW5kZXJPcmRlcjtcblx0fSBlbHNlIHtcblx0XHRyZXR1cm4gYS56IC0gYi56O1xuXHR9XG59KSIsInJ1bnMiOlsxODAwLDIyMDAsMTYwMCwyMjAwLDIwMDAsMjQwMCwxNjAwLDI0MDAsMTQwMCwxODAwLDIyMDAsMTgwMCwyNjAwLDE0MDAsMjIwMCwyMjAwLDI0MDAsMjQwMCwyNDAwLDIwMDAsMTQwMCwxODAwLDI0MDAsMjAwMCwxNjAwLDE4MDAsMjQwMCwyMjAwLDIwMDAsMjIwMCwyNDAwLDE2MDAsMjAwMCwxNjAwLDIyMDAsMjQwMCwyMDAwLDI0MDAsMjQwMCwyNDAwLDIwMDAsMjIwMCwyMDAwLDI0MDAsMTQwMCwyMjAwLDIyMDAsMjQwMCwyNDAwLDI0MDAsMTYwMCwxODAwLDE4MDAsMTgwMCwyMjAwLDI0MDAsMjQwMCwyMDAwLDE4MDAsMjQwMCwxNjAwLDI2MDAsMjIwMCwyNjAwLDIwMDAsMjAwMCwxNjAwLDE4MDAsMjQwMCwyMDAwLDI0MDAsMjAwMCwxMjAwLDIwMDAsMjIwMCwyMDAwLDE2MDAsMjIwMCwxODAwLDIwMDAsMjQwMCwyNDAwLDI0MDAsMjAwMCwyMDAwLDI0MDAsMjQwMCwyMjAwLDI0MDAsMjAwMCwyNDAwLDI0MDAsMTYwMCwxODAwLDE2MDAsMjIwMCwxODAwLDE4MDAsMTYwMCwyNDAwXSwib3BzIjoyMDY0fSx7Im5hbWUiOiJGaW5kIGl0ZW0gMjAwIiwiY29kZSI6ImFycmF5LnNvcnQoKGEsYikgPT4gKGEuZ3JvdXBPcmRlciAtIGIuZ3JvdXBPcmRlcikgfHwgKGEucmVuZGVyT3JkZXIgLSBiLnJlbmRlck9yZGVyKSB8fCAoYS56IC0gYi56KSkiLCJydW5zIjpbMjIwMCwyNDAwLDE2MDAsMjAwMCwyMjAwLDIwMDAsMjAwMCwyNDAwLDIwMDAsMjQwMCwxODAwLDI0MDAsMjQwMCwyNDAwLDE4MDAsMjQwMCwyNDAwLDE4MDAsMjIwMCwyNDAwLDE2MDAsMTgwMCwyMjAwLDIwMDAsMTgwMCwyNDAwLDI0MDAsMjQwMCwyMDAwLDI0MDAsMjQwMCwyMDAwLDIyMDAsMjAwMCwyMDAwLDIwMDAsMjQwMCwyNDAwLDIyMDAsMTYwMCwyNDAwLDE4MDAsMjQwMCwyMDAwLDI2MDAsMjIwMCwyMjAwLDIwMDAsMjYwMCwyNDAwLDIyMDAsMjIwMCwyNDAwLDI0MDAsMTgwMCwyMDAwLDI0MDAsMjQwMCwyMDAwLDI0MDAsMTYwMCwyNDAwLDIyMDAsMjQwMCwyNDAwLDIwMDAsMTQwMCwyNDAwLDE4MDAsMjQwMCwyMjAwLDI0MDAsMjQwMCwyNDAwLDIwMDAsMjQwMCwyNDAwLDE0MDAsMjQwMCwxODAwLDIyMDAsMjAwMCwyNDAwLDI0MDAsMTgwMCwxODAwLDIwMDAsMjQwMCwyMjAwLDIwMDAsMTgwMCwyNDAwLDI0MDAsMjAwMCwyNDAwLDIwMDAsMjIwMCwyMjAwLDIyMDAsMjIwMF0sIm9wcyI6MjE2MH1dLCJ1cGRhdGVkIjoiMjAyMy0wOC0wOFQwNzo0NToxMy40MTRaIn0%3D): If-else chain slightly slower on Chrome and Firefox
- [1 groupOrder, 0 renderOrder, random z](https://perf.link/#eyJpZCI6ImN4c3h5cXY0aXl5IiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6ImNvbnN0IGFycmF5ID0gbmV3IEFycmF5KDEwMDAwKTtcbmZvcihsZXQgaSA9IDA7IGkgPCBhcnJheS5sZW5ndGg7IGkrKykge1xuXHRhcnJheVtpXSA9IHtcblx0XHRncm91cE9yZGVyOiAxLFxuXHRcdHJlbmRlck9yZGVyOiAwLFxuXHRcdHo6IE1hdGgucmFuZG9tKCkgKiAxMDAwMDAgLSA1MDAwMCxcblx0fTtcbn0iLCJ0ZXN0cyI6W3sibmFtZSI6IkZpbmQgaXRlbSAxMDAiLCJjb2RlIjoiYXJyYXkuc29ydCgoYSxiKSA9PiB7XG5cdGlmKGEuZ3JvdXBPcmRlciAhPT0gYi5ncm91cE9yZGVyKSB7XG5cdFx0cmV0dXJuIGEuZ3JvdXBPcmRlciAtIGIuZ3JvdXBPcmRlcjtcblx0fSBlbHNlIGlmKGEucmVuZGVyT3JkZXIgIT09IGIuZ3JvdXBPcmRlcikge1xuXHRcdHJldHVybiBhLnJlbmRlck9yZGVyIC0gYi5yZW5kZXJPcmRlcjtcblx0fSBlbHNlIHtcblx0XHRyZXR1cm4gYS56IC0gYi56O1xuXHR9XG59KSIsInJ1bnMiOls0NzE0LDQyODUsNDAwMCw0NDI4LDQwMDAsMzU3MSwzNTcxLDQyODUsNDU3MSw0Mjg1LDQ1NzEsMjcxNCwzNzE0LDM1NzEsNDI4NSw0NTcxLDQxNDIsNDU3MSw0NTcxLDQ0MjgsMzg1Nyw0MTQyLDQwMDAsNDU3MSw0NDI4LDQ1NzEsNDU3MSw0Mjg1LDQ3MTQsNDU3MSw0NDI4LDQ3MTQsNDQyOCw0NTcxLDQxNDIsMzg1NywzODU3LDQ0MjgsMzg1NywzODU3LDM0MjgsMzcxNCwzNTcxLDQ0MjgsMzg1Nyw0MTQyLDQyODUsNDE0MiwzODU3LDQ3MTQsNDU3MSw0NzE0LDM3MTQsNDE0Miw0MDAwLDM3MTQsMzg1NywzNzE0LDM1NzEsMzg1Nyw0MTQyLDQwMDAsMzg1Nyw0NDI4LDQyODUsNDU3MSwzODU3LDQyODUsNDcxNCw0NDI4LDM3MTQsMzg1NywzODU3LDM0MjgsNDU3MSwzODU3LDM4NTcsMzU3MSwzNTcxLDM3MTQsMzcxNCwzNzE0LDM1NzEsMzQyOCwzNDI4LDQyODUsNDE0MiwzODU3LDQyODUsNDU3MSw0NzE0LDQ0MjgsNDU3MSw0NDI4LDQwMDAsNDU3MSw0Mjg1LDM4NTcsNDAwMCw0MTQyXSwib3BzIjo0MTEyfSx7Im5hbWUiOiJGaW5kIGl0ZW0gMjAwIiwiY29kZSI6ImFycmF5LnNvcnQoKGEsYikgPT4gKGEuZ3JvdXBPcmRlciAtIGIuZ3JvdXBPcmRlcikgfHwgKGEucmVuZGVyT3JkZXIgLSBiLnJlbmRlck9yZGVyKSB8fCAoYS56IC0gYi56KSkiLCJydW5zIjpbMTQyOCwxMjg1LDg1NywxMjg1LDEwMDAsMjg1LDg1NywxMTQyLDcxNCwxMTQyLDE1NzEsMTE0MiwxMjg1LDEyODUsMTE0MiwxMTQyLDU3MSw0MjgsODU3LDU3MSwxMjg1LDEwMDAsODU3LDEyODUsMTE0MiwxMDAwLDcxNCwxNTcxLDg1NywxMTQyLDEwMDAsNzE0LDE1NzEsODU3LDg1Nyw1NzEsMTAwMCwxMDAwLDcxNCwxMTQyLDcxNCwxMTQyLDg1Nyw4NTcsMTAwMCwxMTQyLDExNDIsNzE0LDEyODUsMTI4NSwxNTcxLDg1NywxMDAwLDExNDIsNDI4LDEwMDAsODU3LDEyODUsMTE0MiwxNTcxLDExNDIsODU3LDg1NywxNDI4LDQyOCwxMTQyLDQyOCw1NzEsMTcxNCw4NTcsMTE0Miw4NTcsODU3LDEwMDAsNDI4LDEwMDAsNzE0LDI4NSwyODUsODU3LDQyOCwxMDAwLDg1Nyw3MTQsNTcxLDcxNCwxMjg1LDcxNCw4NTcsMTE0MiwxNzE0LDE0MjgsNzE0LDg1NywxMjg1LDE1NzEsNTcxLDg1NywxMjg1LDExNDJdLCJvcHMiOjk3OH1dLCJ1cGRhdGVkIjoiMjAyMy0wOC0wOFQwNzo0NzozNS42NTBaIn0%3D) If-else completely dominates the compacted expression, both on Chrome and Firefox
- [0 groupOrder, 0 renderOrder, random z](https://perf.link/#eyJpZCI6IjZtMjl5eDI0Mm5hIiwidGl0bGUiOiJGaW5kaW5nIG51bWJlcnMgaW4gYW4gYXJyYXkgb2YgMTAwMCIsImJlZm9yZSI6ImNvbnN0IGFycmF5ID0gbmV3IEFycmF5KDEwMDAwKTtcbmZvcihsZXQgaSA9IDA7IGkgPCBhcnJheS5sZW5ndGg7IGkrKykge1xuXHRhcnJheVtpXSA9IHtcblx0XHRncm91cE9yZGVyOiAwLFxuXHRcdHJlbmRlck9yZGVyOiAwLFxuXHRcdHo6IE1hdGgucmFuZG9tKCkgKiAxMDAwMDAgLSA1MDAwMCxcblx0fTtcbn0iLCJ0ZXN0cyI6W3sibmFtZSI6IkZpbmQgaXRlbSAxMDAiLCJjb2RlIjoiYXJyYXkuc29ydCgoYSxiKSA9PiB7XG5cdGlmKGEuZ3JvdXBPcmRlciAhPT0gYi5ncm91cE9yZGVyKSB7XG5cdFx0cmV0dXJuIGEuZ3JvdXBPcmRlciAtIGIuZ3JvdXBPcmRlcjtcblx0fSBlbHNlIGlmKGEucmVuZGVyT3JkZXIgIT09IGIuZ3JvdXBPcmRlcikge1xuXHRcdHJldHVybiBhLnJlbmRlck9yZGVyIC0gYi5yZW5kZXJPcmRlcjtcblx0fSBlbHNlIHtcblx0XHRyZXR1cm4gYS56IC0gYi56O1xuXHR9XG59KSIsInJ1bnMiOlsxODc1LDEzNzUsMTEyNSwxNTAwLDE1MDAsMTYyNSwxMzc1LDE1MDAsMTYyNSwxMTI1LDE2MjUsMTc1MCwxMDAwLDIwMDAsMjAwMCwxNjI1LDE4NzUsMTc1MCwxMjUwLDEzNzUsMTI1MCwxMjUwLDExMjUsMTUwMCwxMDAwLDg3NSwxMzc1LDExMjUsMTEyNSwyMDAwLDE2MjUsMTUwMCwxNzUwLDE4NzUsMTg3NSwxNzUwLDE2MjUsMTg3NSwxODc1LDExMjUsMTc1MCwxMjUwLDE1MDAsMTg3NSwyMTI1LDE1MDAsMTEyNSwxMzc1LDEyNTAsMTUwMCwxMjUwLDIxMjUsMTg3NSwxODc1LDE2MjUsMTM3NSwxNzUwLDEzNzUsMTAwMCwxNzUwLDE3NTAsMTI1MCw4NzUsMTc1MCwxNzUwLDE2MjUsMTg3NSwxNjI1LDE1MDAsMTUwMCwxNTAwLDE2MjUsMTI1MCwxMjUwLDEyNTAsMTc1MCwyMDAwLDExMjUsMTg3NSwxNTAwLDE4NzUsMTUwMCwxODc1LDE3NTAsMTM3NSwxMzc1LDEzNzUsMTUwMCwyMDAwLDE3NTAsMTI1MCwxNjI1LDE2MjUsMjAwMCwxNTAwLDE1MDAsMTUwMCwyMDAwLDIxMjUsMTM3NV0sIm9wcyI6MTU1MX0seyJuYW1lIjoiRmluZCBpdGVtIDIwMCIsImNvZGUiOiJhcnJheS5zb3J0KChhLGIpID0%2BIChhLmdyb3VwT3JkZXIgLSBiLmdyb3VwT3JkZXIpIHx8IChhLnJlbmRlck9yZGVyIC0gYi5yZW5kZXJPcmRlcikgfHwgKGEueiAtIGIueikpIiwicnVucyI6WzE2MjUsMTM3NSwyMDAwLDE1MDAsMTYyNSwxMzc1LDEzNzUsMTM3NSwxNjI1LDE4NzUsMjAwMCwxNTAwLDExMjUsMjAwMCwxNTAwLDExMjUsMjUwLDE4NzUsMTUwMCwxMzc1LDE3NTAsMTYyNSwxMDAwLDE2MjUsMTg3NSwxMzc1LDYyNSwxMzc1LDExMjUsMTg3NSwxMzc1LDEzNzUsMTYyNSwxNzUwLDEzNzUsMTYyNSwxMzc1LDEyNTAsMTM3NSwxODc1LDE3NTAsMTYyNSwxMzc1LDE2MjUsMTUwMCwxNjI1LDEwMDAsMTAwMCwxMDAwLDE3NTAsMTUwMCwxNjI1LDExMjUsMTM3NSwxMjUwLDEzNzUsMTEyNSwxNzUwLDEzNzUsMjEyNSwxNjI1LDE1MDAsMTc1MCwxNTAwLDE1MDAsMTYyNSwxMzc1LDE1MDAsMTI1MCwxMTI1LDE4NzUsMTYyNSwxNTAwLDExMjUsMTUwMCwxNTAwLDExMjUsMTUwMCwxNjI1LDEzNzUsMTI1MCwzNzUsMTYyNSwxMzc1LDE3NTAsMTYyNSwxODc1LDE4NzUsMTYyNSwxNjI1LDE1MDAsMTYyNSwxNzUwLDE3NTAsMTM3NSwxMzc1LDE2MjUsMTEyNSwzNzUsMTYyNV0sIm9wcyI6MTQ2N31dLCJ1cGRhdGVkIjoiMjAyMy0wOC0wOFQwNzo1MDo1My41NDhaIn0%3D): About equal, if-else slightly faster on Chrome, but slightly slower on Firefox

Result obviously may vary and I don't think the performance of this sorting function would really matter at the end of the day as most VR experiences tend to have relatively low object counts anyway.

Reply:
> Wouldn't it make sense to retain the ID tie-breaker (return a.id - b.id) used by Three.js as well? That should ensure that, everything else being equal, the order is consistent.

That would sort the objects in creation order, rather than DOM order.  My understanding is that returning 0 will leave order as-is, which implies DOM order, which matches current behaviour, is explicable in A-Frame terms, and gives the user a level of control by inserting items earlier in the DOM.

IMO that's superios to object id / creation order.

Reply:
> As for compacting it to a single expression, I personally prefer the current more verbose version. 

Thanks for doing the perf testing, looks to me like the more verbose version is both easier to read, and probably also more performant (I don't think random data is particularly realistic, and even in that case, if-else seems to be only marginally slower).

So I think that weighs in favour of the code as-is.  Which also matches the equivalent THREE.js code.

There may be a case for eliminating the final if test and simply returning `a.z - b.z` though, so I'll make that change.